File: wk6/java/removeAllAdjacentDuplicatesInString.java
// problem link: https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/

class Solution {
    public String removeDuplicates(String s) {
        LinkedList<Character> sta = new LinkedList<>();
        for(int i = 0; i < s.length(); i++) {
            if(sta.isEmpty() || sta.peek() != s.charAt(i))
                sta.push(s.charAt(i));
            else
                sta.pop();
        }
        StringBuilder ret = new StringBuilder();
        while(!sta.isEmpty())
            ret.append(sta.pollLast());
        return ret.toString();
    }
}

File: wk6/cpp/removeAllAdjacentDuplicatesInString.cpp
// problem link: https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/description/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string removeDuplicates(string s) {
        int l = -1;
        for(int r = 0; r < s.length(); r++) {
            if(l == -1 || s[l] != s[r])
                s[++l] = s[r];
            else
                l--;
        }
        s.resize(l + 1);
        return s;
    }
};